File: sdk_src/Server/MobileIMSDKServer_Open/src/net/x52im/mobileimsdk/server/protocal/Protocal.java
/*
 * Copyright (C) 2021  即时通讯网(52im.net) & Jack Jiang.
 * The MobileIMSDK v6.x Project. 
 * All rights reserved.
 * 
 * > Github地址：https://github.com/JackJiang2011/MobileIMSDK
 * > 文档地址：  http://www.52im.net/forum-89-1.html
 * > 技术社区：  http://www.52im.net/
 * > 技术交流群：320837163 (http://www.52im.net/topic-qqgroup.html)
 * > 作者公众号：“【即时通讯技术圈】”，欢迎关注！
 * > 联系作者：  http://www.52im.net/thread-2792-1-1.html
 *  
 * "即时通讯网(52im.net) - 即时通讯开发者社区!" 推荐开源工程。
 * 
 * Protocal.java at 2021-6-29 10:15:35, code by Jack Jiang.
 */
package net.x52im.mobileimsdk.server.protocal;

import java.util.UUID;

import net.x52im.mobileimsdk.server.ServerLauncher;

import com.google.gson.Gson;

public class Protocal
{
	protected boolean bridge = false;
	protected int type = 0;
	protected String dataContent = null;
	protected String from = "-1";
	protected String to = "-1";
	protected String fp = null;
	protected boolean QoS = false;
	protected int typeu = -1;
	
	protected transient int retryCount = 0;
	protected long sm = -1;
	public Protocal(int type, String dataContent, String from, String to)
	{
		this(type, dataContent, from, to, -1);
	}
	
	public Protocal(int type, String dataContent, String from, String to, int typeu)
	{
		this(type, dataContent, from, to, false, null, typeu);
	}

	public Protocal(int type, String dataContent, String from, String to
			, boolean QoS, String fingerPrint)
	{
		this(type, dataContent, from, to, QoS, fingerPrint, -1);
	}
	
	public Protocal(int type, String dataContent, String from, String to
			, boolean QoS, String fingerPrint, int typeu)
	{
		this.type = type;
		this.dataContent = dataContent;
		this.from = from;
		this.to = to;
		this.QoS = QoS;
		this.typeu = typeu;

		if(QoS && fingerPrint == null)
			fp = Protocal.genFingerPrint();
		else
			fp = fingerPrint;
	}
	
	public int getType()
	{
		return this.type;
	}
	public void setType(int type)
	{
		this.type = type;
	}
	
	public String getDataContent()
	{
		return this.dataContent;
	}
	public void setDataContent(String dataContent)
	{
		this.dataContent = dataContent;
	}

	public String getFrom()
	{
		return this.from;
	}
	public void setFrom(String from)
	{
		this.from = from;
	}

	public String getTo()
	{
		return this.to;
	}
	public void setTo(String to)
	{
		this.to = to;
	}
	
	public String getFp()
	{
		return this.fp;
	}

	public int getRetryCount()
	{
		return this.retryCount;
	}
	public void increaseRetryCount()
	{
		this.retryCount += 1;
	}
	
	public boolean isQoS()
	{
		return QoS;
	}
	public void setQoS(boolean qoS)
	{
		this.QoS = qoS;
	}
	
	public boolean isBridge()
	{
		return bridge;
	}
	public void setBridge(boolean bridge)
	{
		this.bridge = bridge;
	}
	
	public int getTypeu()
	{
		return typeu;
	}
	public void setTypeu(int typeu)
	{
		this.typeu = typeu;
	}


	public long getSm()
	{
		return sm;
	}

	public void setSm(long sm)
	{
		this.sm = sm;
	}
	
	public String toGsonString()
	{
		return new Gson().toJson(this);
	}
	
	public byte[] toBytes()
	{
		return CharsetHelper.getBytes(toGsonString());
	}
	
	@Override
	public Object clone()
	{
		Protocal cloneP = new Protocal(this.getType()
				, this.getDataContent(), this.getFrom(), this.getTo(), this.isQoS(), this.getFp());
		cloneP.setBridge(this.bridge); // since 3.0
		cloneP.setTypeu(this.typeu);   // since 3.0
//		cloneP.setCm(this.cm);         // since 6.0
		cloneP.setSm(this.sm);         // since 6.0
		return cloneP;
	}
	
		public static long genServerTimestamp()
	{
		return System.currentTimeMillis();
	}
	public static String genFingerPrint()
	{
		return UUID.randomUUID().toString();
	}
}

File: sdk_src/TCP_Client/MobileIMSDK4i_tcp_Open/MobileIMSDK4i/Core/Protocal/Protocal.m
//  ----------------------------------------------------------------------
//  Copyright (C) 2021  即时通讯网(52im.net) & Jack Jiang.
//  The MobileIMSDK_TCP (MobileIMSDK v6.x TCP版) Project.
//  All rights reserved.
//
//  > Github地址: https://github.com/JackJiang2011/MobileIMSDK
//  > 文档地址:    http://www.52im.net/forum-89-1.html
//  > 技术社区：   http://www.52im.net/
//  > 技术交流群： 215477170 (http://www.52im.net/topic-qqgroup.html)
//  > 作者公众号： “即时通讯技术圈】”，欢迎关注！
//  > 联系作者：   http://www.52im.net/thread-2792-1-1.html
//
//  "即时通讯网(52im.net) - 即时通讯开发者社区!" 推荐开源工程。
//  ----------------------------------------------------------------------

#import "Protocal.h"
#import "ToolKits.h"
#import "RMMapper.h"
#import "CharsetHelper.h"
#import "ProtocalFactory.h"


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
#pragma mark - 私有API
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
@interface Protocal ()

@property (nonatomic, assign) int retryCount;

@end


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
#pragma mark - 本类的代码实现
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

@implementation Protocal

+ (Protocal *) initWithType:(int)type content:(NSString *)dataContent from:(NSString *)from to:(NSString *)to
{
    return [Protocal initWithType:type content:dataContent from:from to:to qos:YES fp:nil tu:-1];
}
+ (Protocal *) initWithType:(int)type content:(NSString *)dataContent from:(NSString *)from to:(NSString *)to tu:(int)typeu
{
    return [Protocal initWithType:type content:dataContent from:from to:to qos:YES fp:nil tu:typeu];
}
+ (Protocal *) initWithType:(int)type content:(NSString *)dataContent from:(NSString *)from to:(NSString *)to qos:(bool)QoS fp:(NSString *)fingerPrint tu:(int)typeu
{
    return [Protocal initWithType:type content:dataContent from:from to:to qos:QoS fp:fingerPrint bg:NO tu:typeu];
}
+ (Protocal *) initWithType:(int)type content:(NSString *)dataContent from:(NSString *)from to:(NSString *)to qos:(bool)QoS fp:(NSString *)fingerPrint bg:(bool)bridge tu:(int)typeu
{
    Protocal *p = [[Protocal alloc] init];
    
    p.type = type;
    p.dataContent = dataContent;
    p.from = from;
    p.to = to;
    p.QoS = QoS;
    p.bridge = bridge;
    p.typeu = typeu;

    if(QoS && fingerPrint == nil)
        p.fp = [Protocal genFingerPrint];
    else
        p.fp = fingerPrint;
    
    return p;
}

- (id)init
{
    if(self = [super init])
    {
        self.type = 0;
        self.from = @"-1";
        self.to = @"-1";
        self.QoS = NO;
        self.bridge = NO;
        self.typeu = -1;
        self.sm = -1;
        
        self.retryCount = 0;
    }
    return self;
}

- (int) getRetryCount
{
    return self.retryCount;
}

- (void) increaseRetryCount
{
    self.retryCount += 1;
}

- (NSString *) toGsonString
{
    return [ToolKits toJSONString:[self toBytes]];
}

- (NSData *) toBytes
{
    NSMutableDictionary *dic = [self toMutableDictionary:YES];
    return [ToolKits toJSONBytesWithDictionary:dic];
}
- (NSMutableDictionary *) toMutableDictionary:(BOOL)deleteRetryCountProperty
{
    NSMutableDictionary *dic = [ToolKits toMutableDictionary:self];
    if(deleteRetryCountProperty)
        [dic removeObjectForKey:@"retryCount"];
    return dic;
}

- (Protocal *) clone
{
    NSMutableDictionary *dic = [self toMutableDictionary:YES];
    Protocal *pepFromJASON = [RMMapper objectWithClass:[Protocal class] fromDictionary:dic];
    return pepFromJASON;
}

+ (NSString *) genFingerPrint
{
    return [ToolKits generateUUID];
}

@end
